File: c++/plots/plots.m
/******************************************************************************
 * Version: 1.0
 * Last modified on: 21 January, 2013 
 * Developers: Michael G. Epitropakis, Xiaodong Li.
 *      email: mge_(AT)_cs_(DOT)_stir_(DOT)_ac_(DOT)_uk 
 *           : xiaodong_(DOT)_li_(AT)_rmit_(DOT)_edu_(DOT)_au 
 * ***************************************************************************/

% F1: Five-Uneven-Peak Trap
figure(1);
data=load('f1.dat');
plot(data(:,1),data(:,2),'b-');
saveas(gcf,'figs/five_uneven_peak_trap.eps','psc2')
saveas(gcf,'figs/five_uneven_peak_trap.png','png')

% F2: Equal Maxima
figure(2);
data=load('f2.dat');
plot(data(:,1),data(:,2),'b-');
saveas(gcf,'figs/equal_maxima.eps','psc2')
saveas(gcf,'figs/equal_maxima.png','png')


% F3: Uneven Decreasing Maxima
figure(3);
data=load('f3.dat');
plot(data(:,1),data(:,2),'b-');
saveas(gcf,'figs/uneven_decreasing_maxima.eps','psc2')
saveas(gcf,'figs/uneven_decreasing_maxima.png','png')

% F4: Himmelblau
figure(4);
[X,Y] = meshgrid(-6:.05:6);
Z=load('f4.dat');
surfc(X,Y,Z, 'FaceColor','interp','FaceLighting','phong','EdgeColor','none');
axis([-6 6 -6 6 -2000 200])
saveas(gcf,'figs/himmelblau.eps','psc2')
saveas(gcf,'figs/himmelblau.png','png')

% F5: Six-Hump Camel Back
figure(5);
[X,Y] = meshgrid([-1.9:0.019:1.9],[-1.1:0.011:1.1]);
Z=load('f5.dat');
surfc(X,Y,Z, 'FaceColor','interp','FaceLighting','phong','EdgeColor','none');
axis([-1.9 1.9 -1.1 1.1 -25 5])
saveas(gcf,'figs/six_hump_camel_back.eps','psc2')
saveas(gcf,'figs/six_hump_camel_back.png','png')


% F6: Shubert
figure(6);
[X,Y] = meshgrid(-10:0.1:10);
Z=load('f6.dat');
surfc(X,Y,Z, 'FaceColor','interp','FaceLighting','phong','EdgeColor','none');
axis([-10 10 -10 10 -300 200])
saveas(gcf,'figs/shubert.eps','psc2')
saveas(gcf,'figs/shubert.png','png')

% F7: Vincent
figure(7);
[X,Y] = meshgrid(0.25:0.05:10);
Z=load('f7.dat');
surfc(X,Y,Z, 'FaceColor','interp','FaceLighting','phong','EdgeColor','none');
axis([0.25 10 0.25 10 -1 1])
saveas(gcf,'figs/vincent.eps','psc2')
saveas(gcf,'figs/vincent.png','png')

% F8: Modified Rastrigin - All Global Optima
figure(8);
[X,Y] = meshgrid(0:0.01:1);
Z=load('f8.dat');
surfc(X,Y,Z, 'FaceColor','interp','FaceLighting','phong','EdgeColor','none');
axis([0 1 0 1 -40 10])
saveas(gcf,'figs/mod_rastrigin_all.eps','psc2')
saveas(gcf,'figs/mod_rastrigin_all.png','png')

% CF1: Composition Function 1
figure(9);
[X,Y] = meshgrid(-5:0.01:5);
Z=load('CF1.dat');
surfc(X,Y,Z, 'FaceColor','interp','FaceLighting','phong','EdgeColor','none');
saveas(gcf,'figs/CF1.eps','psc2')
saveas(gcf,'figs/CF1.png','png')

% CF2: Composition Function 2
figure(10);
[X,Y] = meshgrid(-5:0.01:5);
Z=load('CF2.dat');
surfc(X,Y,Z, 'FaceColor','interp','FaceLighting','phong','EdgeColor','none');
saveas(gcf,'figs/CF2.eps','psc2')
saveas(gcf,'figs/CF2.png','png')

% CF3: Composition Function 3
figure(11);
[X,Y] = meshgrid(-5:0.01:5);
Z=load('CF3.dat');
surfc(X,Y,Z, 'FaceColor','interp','FaceLighting','phong','EdgeColor','none');
saveas(gcf,'figs/CF3.eps','psc2')
saveas(gcf,'figs/CF3.png','png')

% CF4: Composition Function 4
figure(12);
[X,Y] = meshgrid(-5:0.01:5);
Z=load('CF4.dat');
surfc(X,Y,Z, 'FaceColor','interp','FaceLighting','phong','EdgeColor','none');
saveas(gcf,'figs/CF4.eps','psc2')
saveas(gcf,'figs/CF4.png','png')

File: java/CEC2013/src/cec2013/Plots.java
/******************************************************************************
 * Version: 1.0
 * Last modified on: 4 March, 2013 
 * Developers: Michael G. Epitropakis, Xiaodong Li.
 *      email: mge_(AT)_cs_(DOT)_stir_(DOT)_ac_(DOT)_uk 
 *           : xiaodong_(DOT)_li_(AT)_rmit_(DOT)_edu_(DOT)_au 
 * ***************************************************************************/
package cec2013;

import java.io.BufferedWriter;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStreamWriter;

public final class Plots {

	/**
	 * @param args
	 */
	public static void main(String[] args) {
		
	    /**********************************************************************
	     * Function 1 
	     *********************************************************************/
		System.out.println("Preparing files for Function 1...");
		Func f1 = new Funcs.FiveUnevenPeakTrap( new BoundsFn.ConstantBoundsFn( 1, new ClosedInterval.Double(0.0, 30.0) ) );
		Plot1D(f1, "plots/f1.dat", 0.1);
		
	    /**********************************************************************
	     * Function 2 
	     *********************************************************************/
		System.out.println("Preparing files for Function 2...");
		Func f2 = new Funcs.EqualMaxima( new BoundsFn.ConstantBoundsFn( 1, new ClosedInterval.Double(0.0, 1.0) ) );
		Plot1D(f2, "plots/f2.dat", 0.001);
		
	    /**********************************************************************
	     * Function 3 
	     *********************************************************************/
		System.out.println("Preparing files for Function 3...");
		Func f3 = new Funcs.UnevenDecreasingMaxima( new BoundsFn.ConstantBoundsFn( 1, new ClosedInterval.Double(0.0, 1.0) ) );
		Plot1D(f3, "plots/f3.dat", 0.001);
		
	    /**********************************************************************
	     * Function 4 
	     *********************************************************************/
		System.out.println("Preparing files for Function 4...");
		Func f4 = new Funcs.Himmelblau( new BoundsFn.ConstantBoundsFn( 2, new ClosedInterval.Double( -6.0, 6.0 ) ) );
		double [] x = new double[2]; x[0] = x[1] = 0.05;
		Plot2D(f4, "plots/f4.dat", x);

	    /**********************************************************************
	     * Function 5 
	     *********************************************************************/
		System.out.println("Preparing files for Function 5...");
		Func f5 = new Funcs.SixHumpCamelBack( new BoundsFn.ExplicitBoundsFn( 
				new ClosedInterval.Double( -1.9, 1.9 ), 
				new ClosedInterval.Double( -1.1, 1.1 ) ) );
		x[0] = 0.019; x[1] = 0.011;
		Plot2D(f5, "plots/f5.dat", x);

	    /**********************************************************************
	     * Function 6 
	     *********************************************************************/
		System.out.println("Preparing files for Function 6...");
		Func f6 = new Funcs.Shubert( new BoundsFn.ConstantBoundsFn( 2, new ClosedInterval.Double( -10.0, 10.0 ) ) );
		Plot2D(f6, "plots/f6.dat", new double [] {0.1, 0.1});

	    /**********************************************************************
	     * Function 7 
	     *********************************************************************/
		System.out.println("Preparing files for Function 7...");
		Func f7 = new Funcs.Vincent( new BoundsFn.ConstantBoundsFn( 2, new ClosedInterval.Double( 0.25, 10.0 ) ) );
		Plot2D(f7, "plots/f7.dat", new double [] {0.05, 0.05});

		/**********************************************************************
	     * Function 8 
	     *********************************************************************/
		System.out.println("Preparing files for Function 8...");
		Func f8 = new Funcs.ModifiedRastriginAll( new BoundsFn.ConstantBoundsFn( 2, new ClosedInterval.Double( 0.0, 1.0 ) ) );
		Plot2D(f8, "plots/f8.dat", new double [] {0.01, 0.01});

	    /**********************************************************************
	     * Composition functions F9-F10, 2-D data files for CF1-CF4
	     *********************************************************************/
		System.out.println("Preparing files for Composition Function 1...");
		Func cf1 = new Funcs.CF1( new BoundsFn.ConstantBoundsFn( 2, new ClosedInterval.Double( -5, 5 ) ) );
		Plot2D(cf1, "plots/CF1.dat", new double [] {0.01, 0.01});
		
		System.out.println("Preparing files for Composition Function 2...");
		Func cf2 = new Funcs.CF2( new BoundsFn.ConstantBoundsFn( 2, new ClosedInterval.Double( -5, 5 ) ) );
		Plot2D(cf2, "plots/CF2.dat", new double [] {0.01, 0.01});

		System.out.println("Preparing files for Composition Function 3...");
		Func cf3 = new Funcs.CF3( new BoundsFn.ConstantBoundsFn( 2, new ClosedInterval.Double( -5, 5 ) ) );
		Plot2D(cf3, "plots/CF3.dat", new double [] {0.01, 0.01});

		System.out.println("Preparing files for Composition Function 4...");
		Func cf4 = new Funcs.CF4( new BoundsFn.ConstantBoundsFn( 2, new ClosedInterval.Double( -5, 5 ) ) );
		Plot2D(cf4, "plots/CF4.dat", new double [] {0.01, 0.01});

	}//main
	
	public static void Plot1D(Func pFunc, String filename, final double stepsize) {
		try {
			/* Open the file */
			FileOutputStream fos   = new FileOutputStream(filename);
			OutputStreamWriter osw = new OutputStreamWriter(fos);
			BufferedWriter bw      = new BufferedWriter(osw);

			try {
				double []x = new double[pFunc.getDimension()];

				for (x[0]=pFunc.getBounds().get(0).getLower(); 
						x[0]<=pFunc.getBounds().get(0).getUpper(); 
						x[0]=x[0]+stepsize) {
					bw.write(x[0]+"\t"+pFunc.evaluate(x)+"\n");
				}
			} finally {
				/* Close the file */
				bw.close();

			}

		}catch (IOException e) {
			e.printStackTrace();
		}  
	}

	public static void Plot2D(Func pFunc, String filename, final double []stepsize) {
		try {
			/* Open the file */
			FileOutputStream fos   = new FileOutputStream(filename);
			OutputStreamWriter osw = new OutputStreamWriter(fos);
			BufferedWriter bw      = new BufferedWriter(osw);

			try {
				double []x = new double[pFunc.getDimension()];

				for (x[0]=pFunc.getBounds().get(0).getLower(); 
						x[0]<=pFunc.getBounds().get(0).getUpper(); 
						x[0]=x[0]+stepsize[0]) {
					for (x[1]=pFunc.getBounds().get(1).getLower(); 
							x[1]<=pFunc.getBounds().get(1).getUpper(); 
							x[1]=x[1]+stepsize[1]) {
						//System.out.println("0: "+x[0] + " 1: "+x[1]+" f: "+ pFunc.evaluate(x));
						bw.write(pFunc.evaluate(x)+"\t");
					}
					bw.newLine();
				}
			} finally {
				/* Close the file */
				bw.close();

			}

		}catch (IOException e) {
			e.printStackTrace();
		}       
	}
}
